lib/invitations/inviteUserByEmail.ts

```diff
@@ -0,0 +1,208 @@
+  TemplateNotFoundError,
+  UserAlreadyHasAccessError,
+  UserNotFoundError,
+} from "./exceptions";
+import { getTemplateWithAssociatedUsers } from "@lib/templates";
+import { prisma } from "@lib/integration/prismaConnector";
+import { sendEmail } from "@lib/integration/notifyConnector";
+import { inviteToCollaborate } from "@lib/invitations/emailTemplates/inviteToCollaborate";
+import { inviteToForms } from "@lib/invitations/emailTemplates/inviteToForms";
+import { getOrigin } from "@lib/origin";
+import { logMessage } from "@lib/logger";
+import { Invitation } from "@prisma/client";
+import { logEvent } from "@lib/auditLogs";
+import { isValidGovEmail } from "@lib/validation/validation";
+
+/**
+ * Invite someone to the form by email
+ *
+ * @param ability
+ * @param email
+ * @param formId
+ */
+export const inviteUserByEmail = async (
+  ability: UserAbility,
+  email: string,
+  formId: string,
+  message: string
+) => {
+  let invitation;
+
+  const sender = await getUser(ability, ability.userID);
+
+  if (!sender) {
+    throw new UserNotFoundError();
+  }
+
+  const template = await getTemplateWithAssociatedUsers(ability, formId);
+
+  if (!template) {
+    throw new TemplateNotFoundError();
+  }
+
+  // check if user is already associated with the form
+  if (template.users.some((user) => user.email === email)) {
+    throw new UserAlreadyHasAccessError();
+  }
+
+  if (!isValidGovEmail(email)) {
+    throw new InvalidDomainError();
+  }
+
+  // Check if the email domain matches the sender's email domain
+  const senderDomain = sender.email.split("@")[1];
+  const recipientDomain = email.split("@")[1];
+
+  if (senderDomain !== recipientDomain) {
+    throw new MismatchedEmailDomainError();
+  }
+
+  // check if user is already invited to the form
+  const previousInvitation = await _retrieveFormInvitationByEmail(email, formId);
+  if (previousInvitation) {
+    invitation = previousInvitation;
+    // if invitation is expired, delete and recreate
+    if (previousInvitation.expires < new Date()) {
+      // check js dates vs prisma dates (see 2fa)
+      _deleteInvitation(previousInvitation.id);
+      invitation = await _createInvitation(email, formId);
+    }
+
+    // send or resend invitation email
+    _sendInvitationEmail(sender, invitation, message, template.formRecord);
+
+    logEvent(
+      ability.userID,
+      { type: "Form", id: invitation.templateId },
+      "InviteUser",
+      `${sender.id} invited ${invitation.email}`
+    );
+
+    return;
+  }
+
+  // No previous invitation, create one
+  invitation = await _createInvitation(email, formId);
+  _sendInvitationEmail(sender, invitation, message, template.formRecord);
+
+  return;
+};
+
+/**
+ * Retrieve existing invitation for a form by email
+ *
+ * @param email string
+ * @param formId string
+ * @returns Invitation
+ */
+const _retrieveFormInvitationByEmail = async (email: string, formId: string) => {
+  const invitation = await prisma.invitation.findFirst({
```

Comment:
Looking at the retrieve, delete, and create invitation functions I might suggest using a `findMany` here instead.  This way if the system ends up in a weird state where multiple invitations for the same form and user exist we can handle it.
Another way to address this would be to set that the `templateId` and `email` must be unique on a record within the Primsma Schema.
```
model Invitation {
  id           String   @id @default(cuid())
  email        String
  expires      DateTime
  templateId   String
  template     Template @relation(fields: [templateId], references: [id], onDelete: Cascade, onUpdate: NoAction)
  @@unique([email, templateId])
}
```

Reply:
Settled on updating the schema/db: https://github.com/cds-snc/platform-forms-client/pull/4386/commits/4317aaae1db3326119a164a7181f25f6925f9688

Will ensure error handling on creation is in place.